pkg/blueprint/filesystem_customizations.go

```diff
@@ -9,9 +9,147 @@ import (
 )
 
+type PartitioningCustomization struct {
+	MinSize uint64                        `json:"minsize,omitempty" toml:"minsize,omitempty"`
+	Plain   *PlainFilesystemCustomization `json:"plain,omitempty" toml:"plain,omitempty"`
+	LVM     *LVMCustomization             `json:"lvm,omitempty" toml:"lvm,omitempty"`
+	Btrfs   *BtrfsCustomization           `json:"btrfs,omitempty" toml:"btrfs,omitempty"`
+}
+
+type PlainFilesystemCustomization struct {
+	Filesystems []FilesystemCustomization `json:"filesystems,omitempty" toml:"filesystems,omitempty"`
+}
+
 type FilesystemCustomization struct {
-	Mountpoint string `json:"mountpoint,omitempty" toml:"mountpoint,omitempty"`
+	Mountpoint string `json:"mountpoint" toml:"mountpoint"`
 	MinSize    uint64 `json:"minsize,omitempty" toml:"minsize,omitempty"`
+	Label      string `json:"label,omitempty" toml:"label,omitempty"`
+	Type       string `json:"type,omitempty" toml:"type,omitempty"`
+}
+
+type LVMCustomization struct {
+	VolumeGroups []VGCustomization `json:"volume-groups,omitempty" toml:"volume-groups,omitempty"`
+}
+
+type VGCustomization struct {
+	// Volume group name
+	Name string `json:"name" toml:"name"`
+	// Size of the partition that contains the volume group
+	MinSize        uint64            `json:"minsize" toml:"minsize"`
+	LogicalVolumes []LVCustomization `json:"logical-volumes,omitempty" toml:"logical-volumes,omitempty"`
+}
+
+type LVCustomization struct {
+	// Logical volume name
+	Name string `json:"name,omitempty" toml:"name,omitempty"`
+	FilesystemCustomization
+}
+
+type BtrfsCustomization struct {
+	Volumes []BtrfsVolumeCustomization
+}
+
+type BtrfsVolumeCustomization struct {
+	// Size of the btrfs partition/volume.
+	MinSize    uint64 `json:"minsize" toml:"minsize"`
+	Subvolumes []BtrfsSubvolumeCustomization
+}
+
+type BtrfsSubvolumeCustomization struct {
+	Name       string `json:"name" toml:"name"`
+	Mountpoint string `json:"mountpoint" toml:"mountpoint"`
+	// Qgroup in the future??
+}
+
+// Validate checks for customization combinations that are generally not
+// supported or can create conflicts, regardless of specific distro or image
+// type policies.
+func (p *PartitioningCustomization) Validate() error {
+	if p == nil {
+		return nil
+	}
+
+	if p.Btrfs != nil && p.LVM != nil {
+		return fmt.Errorf("btrfs and lvm partitioning cannot be combined")
+	}
+
+	if p.Btrfs != nil && len(p.Btrfs.Volumes) > 1 {
+		return fmt.Errorf("multiple btrfs volumes are not yet supported")
+	}
+
+	if p.LVM != nil && len(p.LVM.VolumeGroups) > 1 {
+		return fmt.Errorf("multiple LVM volume groups are not yet supported")
+	}
+
+	// iterate through everything and look for:
+	// - empty mountpoints (global)
+	// - duplicate mountpoints (global)
+	// - duplicate volume group and logical volume names (lvm)
+	// - duplicate subvolume names (btrfs)
+	// - empty subvolume names (btrfs)
+
+	mountpoints := make(map[string]bool)
+	if p.Plain != nil {
+		for _, mp := range p.Plain.Filesystems {
```

Comment:
We should also validate types here (and for LVM) I think, so we return an error early on garbage filesystem types.

Interesting cases are btrfs and vfat. Btrfs is definitely a no-go under plain, users should just go through the btrfs field. Btrfs under LVM needs the structure to get expanded. While vfat under plain and LVM is sane, it's extremely rare (apart from /boot/efi ofc), so I would probably error on it until someone actually wants it.